File: android/app/src/main/java/market/foodhome/app/navigation/NavigationPolicy.kt
package market.foodhome.app.navigation

import java.net.URI
import java.nio.charset.StandardCharsets

sealed interface NavigationDecision {
    data class Internal(val uri: URI) : NavigationDecision
    data class External(val uri: URI) : NavigationDecision
    data class Blocked(val reason: BlockReason) : NavigationDecision
}

enum class BlockReason {
    EMPTY,
    TOO_LONG,
    MALFORMED,
    FORBIDDEN_SCHEME,
    USER_INFO,
    INVALID_HOST,
    UNEXPECTED_PORT,
    NESTED_URL,
}

class NavigationPolicy(
    private val trustedOrigin: URI,
    private val maxUrlLength: Int = 2048,
) {
    fun classify(rawUrl: String?): NavigationDecision {
        val raw = rawUrl.orEmpty()
        if (raw.isEmpty()) return NavigationDecision.Blocked(BlockReason.EMPTY)
        if (raw.toByteArray(StandardCharsets.UTF_8).size > maxUrlLength) {
            return NavigationDecision.Blocked(BlockReason.TOO_LONG)
        }
        if (raw != raw.trim() || raw.any(::isAmbiguousCharacter)) {
            return NavigationDecision.Blocked(BlockReason.MALFORMED)
        }

        val uri = runCatching { URI(raw).normalize() }.getOrNull()
            ?: return NavigationDecision.Blocked(BlockReason.MALFORMED)
        if (!uri.scheme.equals("https", ignoreCase = true)) {
            return NavigationDecision.Blocked(BlockReason.FORBIDDEN_SCHEME)
        }
        if (uri.userInfo != null) return NavigationDecision.Blocked(BlockReason.USER_INFO)

        val host = uri.host?.lowercase()
            ?: return NavigationDecision.Blocked(BlockReason.INVALID_HOST)
        if (!isUnambiguousHost(host)) {
            return NavigationDecision.Blocked(BlockReason.INVALID_HOST)
        }
        if (containsNestedUrl(uri.rawQuery) || containsNestedUrl(uri.rawFragment)) {
            return NavigationDecision.Blocked(BlockReason.NESTED_URL)
        }

        val trustedHost = trustedOrigin.host.lowercase()
        val port = effectivePort(uri)
        val trustedPort = effectivePort(trustedOrigin)
        if (host == trustedHost && port != trustedPort) {
            return NavigationDecision.Blocked(BlockReason.UNEXPECTED_PORT)
        }

        return if (
            uri.scheme.equals(trustedOrigin.scheme, ignoreCase = true) &&
            host == trustedHost &&
            port == trustedPort
        ) {
            NavigationDecision.Internal(uri)
        } else {
            NavigationDecision.External(uri)
        }
    }

    private fun containsNestedUrl(rawQuery: String?): Boolean {
        var decoded = rawQuery?.takeIf(String::isNotEmpty) ?: return false
        repeat(MAX_DECODE_PASSES + 1) { pass ->
            if (containsForbiddenDestination(decoded)) return true
            if (pass == MAX_DECODE_PASSES) return false
            decoded = percentDecodeOnce(decoded) ?: return true
        }
        return true
    }

    private fun containsForbiddenDestination(value: String): Boolean {
        val normalized = value.lowercase()
        return FORBIDDEN_NESTED_SCHEMES.any(normalized::contains)
    }

    private fun percentDecodeOnce(value: String): String? {
        val result = StringBuilder(value.length)
        var index = 0
        while (index < value.length) {
            if (value[index] != '%') {
                result.append(value[index])
                index += 1
                continue
            }
            if (index + 2 >= value.length) return null
            val high = value[index + 1].digitToIntOrNull(16) ?: return null
            val low = value[index + 2].digitToIntOrNull(16) ?: return null
            result.append(((high shl 4) or low).toChar())
            index += 3
        }
        return result.toString()
    }

    private fun isAmbiguousCharacter(value: Char): Boolean =
        value.isWhitespace() || value.isISOControl() || value == '\\' || value in BIDI_CONTROLS

    private fun isUnambiguousHost(host: String): Boolean =
        !host.endsWith('.') && host.all { it.code in 0x21..0x7E }

    private fun effectivePort(uri: URI): Int = when {
        uri.port != -1 -> uri.port
        uri.scheme.equals("https", ignoreCase = true) -> 443
        else -> -1
    }

    private companion object {
        const val MAX_DECODE_PASSES = 4
        val BIDI_CONTROLS = setOf(
            '\u200E', '\u200F',
            '\u202A', '\u202B', '\u202C', '\u202D', '\u202E',
            '\u2066', '\u2067', '\u2068', '\u2069',
        )
        val FORBIDDEN_NESTED_SCHEMES = listOf(
            "http://",
            "https://",
            "javascript:",
            "data:",
            "file:",
            "intent:",
        )
    }
}

File: ios/FoodHomeApp/Navigation/NavigationPolicy.swift
import Foundation

enum NavigationBlockReason: Equatable {
    case empty
    case tooLong
    case malformed
    case forbiddenScheme
    case userInfo
    case invalidHost
    case unexpectedPort
    case nestedURL
}

enum NavigationDecision: Equatable {
    case `internal`(URL)
    case external(URL)
    case blocked(NavigationBlockReason)
}

struct NavigationPolicy {
    let trustedOrigin: URL
    var maxURLLength = 2_048

    func classify(_ rawURL: String?) -> NavigationDecision {
        let raw = rawURL ?? ""
        guard !raw.isEmpty else { return .blocked(.empty) }
        guard raw.utf8.count <= maxURLLength else { return .blocked(.tooLong) }
        guard raw == raw.trimmingCharacters(in: .whitespacesAndNewlines),
              !raw.unicodeScalars.contains(where: Self.isAmbiguousCharacter)
        else {
            return .blocked(.malformed)
        }
        guard let url = URL(string: raw),
              let components = URLComponents(url: url, resolvingAgainstBaseURL: false)
        else {
            return .blocked(.malformed)
        }
        guard components.scheme?.lowercased() == "https" else {
            return .blocked(.forbiddenScheme)
        }
        guard components.user == nil, components.password == nil else {
            return .blocked(.userInfo)
        }
        guard let host = components.host?.lowercased() else {
            return .blocked(.invalidHost)
        }
        guard Self.isUnambiguousHost(host) else {
            return .blocked(.invalidHost)
        }
        guard !containsNestedURL(components.percentEncodedQuery),
              !containsNestedURL(components.percentEncodedFragment)
        else {
            return .blocked(.nestedURL)
        }
        guard let trusted = URLComponents(url: trustedOrigin, resolvingAgainstBaseURL: false),
              let trustedHost = trusted.host?.lowercased()
        else {
            return .blocked(.malformed)
        }

        if host == trustedHost && effectivePort(components) != effectivePort(trusted) {
            return .blocked(.unexpectedPort)
        }
        if components.scheme?.lowercased() == trusted.scheme?.lowercased(),
           host == trustedHost,
           effectivePort(components) == effectivePort(trusted) {
            return .internal(url)
        }
        return .external(url)
    }

    private func containsNestedURL(_ rawQuery: String?) -> Bool {
        guard var decoded = rawQuery, !decoded.isEmpty else { return false }
        for pass in 0...Self.maximumDecodePasses {
            if Self.containsForbiddenDestination(decoded) { return true }
            if pass == Self.maximumDecodePasses { return false }
            guard let next = Self.percentDecodeOnce(decoded) else { return true }
            decoded = next
        }
        return true
    }

    private static func containsForbiddenDestination(_ value: String) -> Bool {
        let normalized = value.lowercased()
        return forbiddenNestedSchemes.contains { normalized.contains($0) }
    }

    private static func percentDecodeOnce(_ value: String) -> String? {
        let input = Array(value.utf8)
        var output: [UInt8] = []
        output.reserveCapacity(input.count)
        var index = 0
        while index < input.count {
            guard input[index] == 0x25 else {
                output.append(input[index])
                index += 1
                continue
            }
            guard index + 2 < input.count,
                  let high = hexValue(input[index + 1]),
                  let low = hexValue(input[index + 2])
            else {
                return nil
            }
            output.append((high << 4) | low)
            index += 3
        }
        return String(bytes: output, encoding: .utf8)
    }

    private static func hexValue(_ byte: UInt8) -> UInt8? {
        switch byte {
        case 0x30...0x39: return byte - 0x30
        case 0x41...0x46: return byte - 0x41 + 10
        case 0x61...0x66: return byte - 0x61 + 10
        default: return nil
        }
    }

    private static func isAmbiguousCharacter(_ scalar: Unicode.Scalar) -> Bool {
        CharacterSet.whitespacesAndNewlines.contains(scalar) ||
            CharacterSet.controlCharacters.contains(scalar) ||
            scalar.value == 0x5C ||
            bidiControls.contains(scalar.value)
    }

    private static func isUnambiguousHost(_ host: String) -> Bool {
        !host.hasSuffix(".") && host.unicodeScalars.allSatisfy {
            $0.value > 0x20 && $0.value <= 0x7F
        }
    }

    private func effectivePort(_ components: URLComponents) -> Int {
        components.port ?? (components.scheme?.lowercased() == "https" ? 443 : -1)
    }

    private static let maximumDecodePasses = 4
    private static let bidiControls: Set<UInt32> = [
        0x200E, 0x200F,
        0x202A, 0x202B, 0x202C, 0x202D, 0x202E,
        0x2066, 0x2067, 0x2068, 0x2069,
    ]
    private static let forbiddenNestedSchemes = [
        "http://", "https://", "javascript:", "data:", "file:", "intent:",
    ]
}
